File: ith_order_statics.cpp
#include<bits/stdc++.h>
using namespace std;

int partition(int A[],int p,int q){
    int x = A[q], i = p;
    for (int j = p; j <= q - 1; j++)
    {
        if (A[j] <= x)
        {
            swap(A[i], A[j]);
            i++;
        }
    }
    swap(A[i], A[q]);
    return i;
}

int randomize_partition(int A[],int p,int q){
    int id=rand()%(q-p+1)+p;
    swap(A[id],A[q]);
    return partition(A,p,q);
}

int randomize_select(int A[],int p,int q,int i){
    if(p==q and i>1) {
        cout<<"error"<<endl;
        return -1;
    }

    else{
        int r=randomize_partition(A,p,q);
        int k=r-p+1;
        if(k==i) return A[r];
        else if(k >i){
            return randomize_select(A,p,r-1,i);
        }
        return randomize_select(A,r+1,q,i-k);
    }
}

int main(){
    int A[]={8,7,9,64,8,6,5,4,2,33};
    int n=sizeof(A)/sizeof(A[0]);

    int ans=randomize_select(A,0,n-1,5);
    cout<<ans<<endl;
    return 0;
}

File: ith_order_statics.py
from random import randint

def swap(A, i, j):
    t = A[i]
    A[i] = A[j]
    A[j] = t

def partition(A,p,q):
    pivot=A[q]
    i=p
    for j in range(p,q):
        if(A[j]<=pivot):
            swap(A,i,j)
            i+=1
    swap(A,i,q)
    return i

def randomised_partition(A, p, q): 
    id = randint(p, q)

    swap(A, id, q)
    return partition(A,p,q)


def randomised_select(A, p, q, i):
    if p == q and i > 1:
        print("error")
        return

    r = randomised_partition(A, p, q)
    k = r-p+1
    if(k == i):
        return A[r]
    elif i < k:
        return randomised_select(A, p, r-1, i)

    return randomised_select(A, r+1, q, i-k)

if __name__=="__main__":
    A= [2, 1, 7, 9, 6, 4, 8, 3]
    ans=randomised_select(A,0,len(A)-1,5)
    print(ans)
